add module field to quickfix dict entry

coot opened the pull request:
Can be used to show module name instead of file name in
quickfix/location list.

I got one problem though with test64 (regexp nfa engine - in log_subexpr).  For me this seems unrelated, but I only get it with this patch so I must be wrong:
```
#0  0x00007ffff4f0b41b in vfprintf () from /lib64/libc.so.6
#1  0x00007ffff4f11aff in fprintf () from /lib64/libc.so.6
#2  0x0000000000570d46 in log_subexpr (sub=0x7fffffffbb40) at regexp_nfa.c:3936
#3  0x0000000000570ba5 in log_subsexpr (subs=0x7fffffffbb40) at regexp_nfa.c:3909
#4  0x00000000005743f6 in nfa_regmatch (prog=0x9a3a90, start=0x9a3bf8, submatch=0x7fffffffb950, m=0x7fffffffbb40) at regexp_nfa.c:5806
#5  0x000000000057336e in recursive_regmatch (state=0x9a3c70, pim=0x0, prog=0x9a3a90, submatch=0x7fffffffb950, m=0x7fffffffbb40, listids=0x7fffffffb620) at regexp_nfa.c:5223
#6  0x0000000000574570 in nfa_regmatch (prog=0x9a3a90, start=0x9a3d60, submatch=0x7fffffffb950, m=0x7fffffffbb40) at regexp_nfa.c:5850
#7  0x00000000005775f4 in nfa_regtry (prog=0x9a3a90, col=1, tm=0x0) at regexp_nfa.c:6994
#8  0x0000000000577dd6 in nfa_regexec_both (line=0x99b3d0 " if then else", startcol=0, tm=0x0) at regexp_nfa.c:7186
#9  0x00000000005781eb in nfa_regexec_nl (rmp=0x7fffffffbf70, line=0x99b3d0 " if then else", col=0, line_lbr=1) at regexp_nfa.c:7345
#10 0x000000000057863a in vim_regexec_both (rmp=0x7fffffffbf70, line=0x99b3d0 " if then else", col=0, nl=1) at regexp.c:8217
#11 0x00000000005788d6 in vim_regexec_nl (rmp=0x7fffffffbf70, line=0x99b3d0 " if then else", col=0) at regexp.c:8291
#12 0x0000000000453705 in find_some_match (argvars=0x7fffffffc450, rettv=0x7fffffffcb00, type=3) at evalfunc.c:7313
#13 0x000000000045418f in f_matchlist (argvars=0x7fffffffc450, rettv=0x7fffffffcb00) at evalfunc.c:7600
#14 0x0000000000448764 in call_internal_func (name=0x9a0db0 "matchlist", argcount=2, argvars=0x7fffffffc450, rettv=0x7fffffffcb00) at evalfunc.c:991
#15 0x00000000005f2fd2 in call_func (funcname=0x9a0790 "matchlist(text, pat)", len=9, rettv=0x7fffffffcb00, argcount_in=2, argvars_in=0x7fffffffc450, argv_func=0x0, firstline=1464, 
    lastline=1464, doesrange=0x7fffffffc604, evaluate=1, partial=0x0, selfdict_in=0x0) at userfunc.c:1446
#16 0x00000000005f0dd3 in get_func_tv (name=0x9a0790 "matchlist(text, pat)", len=9, rettv=0x7fffffffcb00, arg=0x7fffffffcaa0, firstline=1464, lastline=1464, doesrange=0x7fffffffc604, 
    evaluate=1, partial=0x0, selfdict=0x0) at userfunc.c:455
#17 0x000000000043d933 in eval7 (arg=0x7fffffffcaa0, rettv=0x7fffffffcb00, evaluate=1, want_string=0) at eval.c:4332
#18 0x000000000043d129 in eval6 (arg=0x7fffffffcaa0, rettv=0x7fffffffcb00, evaluate=1, want_string=0) at eval.c:3969
#19 0x000000000043cc08 in eval5 (arg=0x7fffffffcaa0, rettv=0x7fffffffcb00, evaluate=1) at eval.c:3785
#20 0x000000000043bec1 in eval4 (arg=0x7fffffffcaa0, rettv=0x7fffffffcb00, evaluate=1) at eval.c:3484
#21 0x000000000043bceb in eval3 (arg=0x7fffffffcaa0, rettv=0x7fffffffcb00, evaluate=1) at eval.c:3401
#22 0x000000000043bb41 in eval2 (arg=0x7fffffffcaa0, rettv=0x7fffffffcb00, evaluate=1) at eval.c:3333
#23 0x000000000043b955 in eval1 (arg=0x7fffffffcaa0, rettv=0x7fffffffcb00, evaluate=1) at eval.c:3261
#24 0x000000000043b88b in eval0 (arg=0x98237f "matchlist(text, pat)", rettv=0x7fffffffcb00, nextcmd=0x7fffffffcbe8, evaluate=1) at eval.c:3221
#25 0x000000000043764e in ex_let (eap=0x7fffffffcbe0) at eval.c:1228
#26 0x000000000047b4d6 in do_one_cmd (cmdlinep=0x7fffffffce00, sourcing=0, cstack=0x7fffffffcef0, fgetline=0x478a8f <get_loop_line>, cookie=0x7fffffffce90) at ex_docmd.c:2947
#27 0x0000000000477f15 in do_cmdline (cmdline=0x0, fgetline=0x4939e2 <getexline>, cookie=0x0, flags=0) at ex_docmd.c:1086
#28 0x00000000005100c8 in nv_colon (cap=0x7fffffffd470) at normal.c:5403
#29 0x0000000000508628 in normal_cmd (oap=0x7fffffffd500, toplevel=1) at normal.c:1150
#30 0x0000000000634b8c in main_loop (cmdwin=0, noexmode=0) at main.c:1343
#31 0x00000000006342e0 in vim_main2 () at main.c:896
#32 0x00000000006339e4 in main (argc=11, argv=0x7fffffffd6f8) at main.c:419
```

This is triggered by this test:
```
:call add(tl, [2, 'if \(\(then\)\@!.\)*$', ' if then else'])
```

Anybody willing to help on this?

coot commented:
ups, probably an issue would be more appropriate...

LemonBoy commented on src/quickfix.c:
```diff
@@ -2561,19 +2574,22 @@ qf_list(exarg_T *eap)
 	    if (got_int)
 		break;
 
-	    fname = NULL;
```
You forgot to initialize `fname`

coot replied:
Thanks @LemonBoy for catching the errors.

LemonBoy commented on src/quickfix.c:
```diff
@@ -2561,19 +2574,22 @@ qf_list(exarg_T *eap)
-		fname = buf->b_fname;
-		if (qfp->qf_type == 1)	/* :helpgrep */
-		    fname = gettail(fname);
+	    if (qfp->qf_module != NULL && STRCMP(qfp->qf_module, "") != 0)
```
The `STRCMP` can be replaced by `*qfp->qf_module != NUL`

coot commented:
@brammool I think this is ready for a review.  Here's a brief summary of the PR:

* add `module` entry in quick fix dictionary
* `module` is optional and it will show up in quick fix window instead of file name
* new errorformat sequence: `%o` (`%m` is already taken) for reading a `.\+` into module
* includes a test which test the reading `%o` escape sequence, if it is parsed correctly into `getqflist()` and if it is shown as intended in quickfix window.

For some languages the file paths can be really long.  For example in `PureScrtip` we often have locations as long as:
```
./src/bower_components/purescript-free/src/Control/Comonad/Cofree.purs
```
And it might get even longer when it expands to a fullpath.  This might be helpful for other languages where there is a clear mapping between file system and module name space (Haskell, Python, NodeJs, etc).

brammool commented:
Marcin Szamotulski wrote:

> @brammool I think this is ready for a review.  Here's a brief summary of the PR:
> 
> * add `module` entry in quick fix dictionary
> * `module` is optional and it will show up in quick fix window instead of file name
> * new errorformat sequence: `%o` (`%m` is already taken) for reading a `.\+` into module
> * includes a test which test the reading `%o` escape sequence, if it is parsed correctly into `getqflist()` and if it is shown as intended in quickfix window.
> 
> For some languages the file paths can be really long.  For example in `PureScrtip` we often have locations as long as:
> ```
> ./src/bower_components/purescript-free/src/Control/Comonad/Cofree.purs
> ```
> And it might get even longer when it expands to a fullpath.  This
> might be helpful for other languages where there is a clear mapping
> between file system and module name space (Haskell, Python, NodeJs,
> etc).

I'll await remarks from Yegappan.

-- 
How To Keep A Healthy Level Of Insanity:
14. Put mosquito netting around your work area. Play a tape of jungle
    sounds all day.

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

yegappan commented on runtime/doc/quickfix.txt:
```diff
@@ -1105,6 +1106,9 @@ conversion can be used to locate lines without a line number in the error
 
+The "%o" conversion will set up module name in quickfix list.  If present it
+will be used in quickfix error window instead of the filename.
+
```
You may want to mention that the module name is only used for display purposes
and the file name is used when jumping to the file.

yegappan commented on src/quickfix.c:
```diff
@@ -1010,6 +1014,21 @@ qf_parse_line(
+		if (regmatch.startp[i] == NULL)
+		    continue;
+		len = (int)(regmatch.endp[i] - regmatch.startp[i]);
+		if (len >= fields->errmsglen)
```
This is not correct. The length of the module string is CMDBUFFSIZE.
If "len" is greater than CMDBUFFSIZE, then len should be truncated.
You can copy the code from the block that handles "%s".

yegappan commented on src/quickfix.c:
```diff
@@ -1186,10 +1205,11 @@ qf_init_ext(
 	convert_setup(&state.vc, enc, p_enc);
 #endif
     fields.namebuf = alloc_id(CMDBUFFSIZE + 1, aid_qf_namebuf);
+    fields.module = alloc_id(CMDBUFFSIZE + 1, aid_qf_namebuf);
```
The aid_qf_namebuf identifier cannot be used for this. You need to create
a new alloc identifier for module name (aid_qf_module). A new enum should
be added to alloc.h.

yegappan commented on src/quickfix.c:
```diff
@@ -2568,24 +2601,29 @@ qf_list(exarg_T *eap)
     for (i = 1; !got_int && i <= qi->qf_lists[qi->qf_curlist].qf_count; )
     {
 	if ((qfp->qf_valid || all) && idx1 <= i && i <= idx2)
+
```
This change is not needed.

yegappan commented on src/testdir/test_quickfix.vim:
```diff
@@ -1085,6 +1085,16 @@ func Test_efm2()
+  let l = getqflist()
+  call assert_equal(1, len(l), string(l))
+  call assert_equal('Language.PureScript.Types', l[0].module)
+  copen
```
You may want to add a test to jump to the entry to make sure that the filename
is used.

coot commented:
@brammool is this in the todo list? I fixed all the things @yegappan had asked for.

brammool commented:
Yes, but it slipped down.  I'll move it up

coot commented:
Thanks!

coot commented:
Purescript vim integration is waiting for this to land in vim.

coot commented:
@brammool did this PR slipped further down?

brammool commented:
Marcin Szamotulski wrote:

> @brammool did this PR slipped further down?

It's in the list.  There was some confusion about what exactly to
include.  I still have to look at what this actually adds and what the
other patch adds on top of it.

-- 
hundred-and-one symptoms of being an internet addict:
182. You may not know what is happening in the world, but you know
     every bit of net-gossip there is.

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

coot commented:
This PR includes:
* add module name to quickfix dictionary
* use module name rather than filename for the `:clist` and `:llist` commands
* adds `%o` to parse a module from an error message

I closed the other [PR](https://github.com/vim/vim/pull/2327) which also extended the support for `filter` command.  There is some overlap, I will rebase it and reopen it when this will get merged first.

I've been running them for a few months without a problem, but there are parts I haven't used extensively (like the parsing), since in PureScript error messages do not come from compiler output, but already come in parsed format as a blob of json.

Commit:
Added module name to quickfix dict

* setqflist / getqflist / clist
* added %o error format for parsing module names in error messages

Commit:
Added tests for quickfix error list module names

* Parsing %o errorformat
* Using module name in quickfix entries

coot commented:
@brammool I updated this patch and also the other branch #2327. They are now rebased agains the current master. Maintaining them against the refactoring in `quickfix.c` is a bit costly, so I hope it can get merged. Maybe there is sombody else that could review it, if it would speeds up the process.
